colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,290 @@
+                        state["exp_avg"] = torch.zeros_like(grad)
+                    if factored:
+                        if sharding_spec.sharding_sequence[0] == 'R': # Col Parallel
+                            state["exp_avg_sq_row"] = torch.zeros(param_height).to(grad)  # [H:4096]
+                            state["exp_avg_sq_col"] = torch.zeros(param_width_parallel).to(grad)  # [W/N:2048]
+                        
+                        if sharding_spec.sharding_sequence[-1] == 'R': # Row Parallel
+                            state["exp_avg_sq_row"] = torch.zeros(param_height_parallel).to(grad)  # [H/N:2048]
+                            state["exp_avg_sq_col"] = torch.zeros(param_width).to(grad)  # [W:4096]
+                            
+                    else:
+                        state["exp_avg_sq"] = torch.zeros_like(grad)
+                    state["RMS"] = 0
+                else:
+                    if use_first_moment:
+                        state["exp_avg"] = state["exp_avg"].to(grad)
+                    if factored:
+                        state["exp_avg_sq_row"] = state["exp_avg_sq_row"].to(grad)
+                        state["exp_avg_sq_col"] = state["exp_avg_sq_col"].to(grad)
+                    else:
+                        state["exp_avg_sq"] = state["exp_avg_sq"].to(grad)
+
+                p_data_fp32 = p.to(self.localRank)
+                if p.dtype in {torch.float16, torch.bfloat16}:
+                    p_data_fp32 = p_data_fp32.float()
+                state["step"] += 1
+                lr = self._get_lr(group, state)
+                beta2t = 1.0 - math.pow(state["step"], group["decay_rate"])
+                update = (grad**2) + group["eps"][0]
+                if factored:  
+                    # ==============================
+                    # First Dim is R, Last Dim is S{} means split dim -1  ---> 
+                    # Coloum Parallel ---> sq_row need Do (col) Reduce
+                    # ==============================
+                    if sharding_spec.sharding_sequence[0] == 'R': 
+                        update_reshape = update.view(-1, param_width_parallel)
+                        grad_reshape = grad.view(-1, param_width_parallel)
+                        exp_avg_sq_row = state["exp_avg_sq_row"] # [H]
+                        exp_avg_sq_col = state["exp_avg_sq_col"] # [W/N]
+                        exp_avg_sq_row.mul_(beta2t).add_(update_reshape.mean(dim=-1), alpha=(1.0 - beta2t))
+                        exp_avg_sq_col.mul_(beta2t).add_(update_reshape.mean(dim=-2), alpha=(1.0 - beta2t))
+                        dist.all_reduce(exp_avg_sq_row, group=self.tensor_parallel_group)
+                        exp_avg_sq_row.div_(self.tensor_parallel_size)
+                        update_reshape = self._approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col)
+                        update_reshape.mul_(grad_reshape)
+                        update = update_reshape.flatten()
```

Comment:
Why?

Reply:
> Why?
To calculate row mean, we need a allreduce, if current param is a column parallel param;
![fig1](https://github.com/hpcaitech/ColossalAI/assets/54985467/7b9f3864-b8f5-45d8-a9ba-dfae1cd5df91)

Reply:
What's the purpose of `flatten()`?

Reply:
> What's the purpose of `flatten()`?

Cause I view the "update_reshape", then I need to perform a flatten().

Reply:
`flatten()` should not be default behavior, as we may not use zero. You may use `view()`

Reply:
You should add readme.md or comments to illustrate your design

Reply:
> `flatten()` should not be default behavior, as we may not use zero. You may use `view()`


fixed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/ce89a849882e845afcc3e99f4a2bdfd9c40f4ee6;
Use view(-1), to replace flatten();